File: src/main/java/ru/octol1ttle/flightassistant/hud/impl/RadarAltitudeDisplay.java
package ru.octol1ttle.flightassistant.hud.impl;

import java.awt.Color;
import net.minecraft.client.font.TextRenderer;
import net.minecraft.client.gui.DrawContext;
import net.minecraft.text.Text;
import net.minecraft.util.math.MathHelper;
import ru.octol1ttle.flightassistant.Dimensions;
import ru.octol1ttle.flightassistant.DrawHelper;
import ru.octol1ttle.flightassistant.computers.impl.AirDataComputer;
import ru.octol1ttle.flightassistant.computers.impl.navigation.FlightPlanner;
import ru.octol1ttle.flightassistant.config.FAConfig;
import ru.octol1ttle.flightassistant.hud.api.IHudDisplay;
import ru.octol1ttle.flightassistant.registries.ComputerRegistry;

public class RadarAltitudeDisplay implements IHudDisplay {
    private final Dimensions dim;
    private final AirDataComputer data = ComputerRegistry.resolve(AirDataComputer.class);
    private final FlightPlanner plan = ComputerRegistry.resolve(FlightPlanner.class);

    public RadarAltitudeDisplay(Dimensions dim) {
        this.dim = dim;
    }

    @Override
    public void render(DrawContext context, TextRenderer textRenderer) {
        if (!FAConfig.indicator().showGroundAltitude) {
            return;
        }
        if (!data.isCurrentChunkLoaded) {
            renderFaulted(context, textRenderer);
            return;
        }

        int bottom = dim.bFrame;
        int xAltText = dim.rFrame + 7;
        int safeLevel = data.groundLevel == data.voidLevel() ? data.voidLevel() + 16 : MathHelper.ceil(data.groundLevel);
        Color color = getAltitudeColor(safeLevel, data.altitude());

        DrawHelper.drawText(textRenderer, context, Text.translatable(data.groundLevel == data.voidLevel() ? "short.flightassistant.void_level" : "short.flightassistant.ground_level"), xAltText - 10, bottom, color);
        DrawHelper.drawText(textRenderer, context, DrawHelper.asText("%d", MathHelper.floor(data.heightAboveGround())), xAltText, bottom, color);
        DrawHelper.drawBorder(context, xAltText - 2, bottom - 2, 28, color);
    }

    private Color getAltitudeColor(int safeLevel, float altitude) {
        if (altitude <= safeLevel) {
            return FAConfig.indicator().warningColor;
        }

        Integer minimums = plan.getMinimums(data.groundLevel);
        if (minimums != null && altitude <= minimums) {
            return FAConfig.indicator().cautionColor;
        }

        return FAConfig.indicator().frameColor;
    }

    @Override
    public void renderFaulted(DrawContext context, TextRenderer textRenderer) {
        DrawHelper.drawText(textRenderer, context, Text.translatable("short.flightassistant.radar"), dim.rFrame - 3, dim.bFrame, FAConfig.indicator().warningColor);
    }

    @Override
    public String getId() {
        return "radar";
    }
}

File: src/client/kotlin/ru/octol1ttle/flightassistant/impl/display/RadarAltitudeDisplay.kt
package ru.octol1ttle.flightassistant.impl.display

import kotlin.math.roundToInt
import net.minecraft.client.gui.DrawContext
import net.minecraft.text.*
import net.minecraft.util.Identifier
import ru.octol1ttle.flightassistant.FlightAssistant
import ru.octol1ttle.flightassistant.api.computer.ComputerAccess
import ru.octol1ttle.flightassistant.api.display.*
import ru.octol1ttle.flightassistant.api.util.*
import ru.octol1ttle.flightassistant.config.FAConfig

class RadarAltitudeDisplay : Display() {
    override fun enabled(): Boolean {
        return FAConfig.display.showAltitudeReading || FAConfig.display.showAltitudeScale
    }

    override fun render(drawContext: DrawContext, computers: ComputerAccess) {
        val groundLevel: Double? = computers.data.groundLevel
        if (groundLevel != null && groundLevel > computers.data.position.y) {
            renderFaulted(drawContext)
            return
        }

        with(drawContext) {
            val x: Int = HudFrame.right - 1
            val y: Int = HudFrame.bottom + 2

            val altType: MutableText
            val altString: String
            val color: Int
            if (groundLevel != null) {
                altType = Text.translatable("short.flightassistant.ground")
                altString = (computers.data.position.y - groundLevel).roundToInt().toString()
                color = primaryColor
            } else {
                altType = Text.translatable("short.flightassistant.void")
                val altValue: Int = (computers.data.position.y - computers.data.voidLevel).roundToInt()
                altString = altValue.toString()
                color = if (altValue <= 0) warningColor else cautionColor
            }
            val xOffset: Int = getTextWidth(altType) + 1

            drawText(altType, x, y + 2, color)
            drawBorder(x + xOffset, y, getTextWidth(altString) + 5, 11, color)
            drawText(altString, x + xOffset + 3, y + 2, color)
        }
    }

    override fun renderFaulted(drawContext: DrawContext) {
        with(drawContext) {
            drawText(
                Text.translatable("short.flightassistant.radar_altitude"),
                HudFrame.right - 1,
                HudFrame.bottom + 4,
                warningColor
            )
        }
    }

    companion object {
        val ID: Identifier = FlightAssistant.id("radar_altitude")
    }
}
